port/cpl_vsisimple.cpp

```diff
@@ -1371,20 +1371,57 @@ GIntBig CPLGetPhysicalRAM(void)
         return 0;
     GIntBig nVal = static_cast<GIntBig>(nPhysPages) * nPageSize;
 
-    // In a Docker container the memory might be limited
-    // If no limitation, on 64 bit, 9223372036854771712 is returned.
-    FILE *f = fopen("/sys/fs/cgroup/memory/memory.limit_in_bytes", "rb");
-    if (f)
+#ifdef __linux
+    char szMemoryPath[256];
+    szMemoryPath[0] = 0;
     {
-        char szBuffer[32];
-        const int nRead =
-            static_cast<int>(fread(szBuffer, 1, sizeof(szBuffer) - 1, f));
-        szBuffer[nRead] = 0;
+        FILE *f = fopen("/proc/self/cgroup", "rb");
+        char szLine[256];
+        // Find line like "6:memory:/user.slice/user-1000.slice/user@1000.service"
+        // and store "/user.slice/user-1000.slice/user@1000.service" in szMemoryPath
+        while (fgets(szLine, sizeof(szLine), f))
+        {
+            const char *pszMemory = strstr(szLine, ":memory:");
+            if (pszMemory)
+            {
+                strcpy(szMemoryPath, pszMemory + strlen(":memory:"));
+                char *pszEOL = strchr(szMemoryPath, '\n');
+                if (pszEOL)
+                    *pszEOL = '\0';
+                break;
+            }
+        }
         fclose(f);
-        const GUIntBig nLimit = CPLScanUIntBig(szBuffer, nRead);
-        nVal =
-            static_cast<GIntBig>(std::min(static_cast<GUIntBig>(nVal), nLimit));
     }
+    if (szMemoryPath[0])
+    {
+        char szFilename[256 + 64];
+        // Read memory.limit_in_byte in the whole szMemoryPath hierarchy
+        while (true)
+        {
+            snprintf(szFilename, sizeof(szFilename),
```

Comment:
I think this only works with the `cgroupv1` hierarchy, but most distros have switched or are switching to `cgroupv2` by now. See https://chrisdown.name/talks/cgroupv2/cgroupv2-fosdem.pdf for more details.

On my distro, I can spawn a new cgroup with `systemd-run --scope -p MemoryMax=250M -p MemoryHigh=200M bash`, then:

```sh
cat /proc/self/cgroup 
0::/system.slice/run-u4260.scope
$ ls /sys/fs/cgroup/system.slice/run-u4260.scope
cgroup.controllers      cgroup.threads         cpuset.mems.effective     hugetlb.1GB.numa_stat     io.bfq.weight   memory.events        memory.reclaim        misc.max
cgroup.events           cgroup.type            cpu.stat                  hugetlb.1GB.rsvd.current  io.latency      memory.events.local  memory.stat           pids.current
cgroup.freeze           cpu.idle               cpu.uclamp.max            hugetlb.1GB.rsvd.max      io.low          memory.high          memory.swap.current   pids.events
cgroup.kill             cpu.max                cpu.uclamp.min            hugetlb.2MB.current       io.max          memory.low           memory.swap.events    pids.max
cgroup.max.depth        cpu.max.burst          cpu.weight                hugetlb.2MB.events        io.pressure     memory.max           memory.swap.high      pids.peak
cgroup.max.descendants  cpu.pressure           cpu.weight.nice           hugetlb.2MB.events.local  io.prio.class   memory.min           memory.swap.max       rdma.current
cgroup.pressure         cpuset.cpus            hugetlb.1GB.current       hugetlb.2MB.max           io.stat         memory.numa_stat     memory.zswap.current  rdma.max
cgroup.procs            cpuset.cpus.effective  hugetlb.1GB.events        hugetlb.2MB.numa_stat     io.weight       memory.oom.group     memory.zswap.max
cgroup.stat             cpuset.cpus.partition  hugetlb.1GB.events.local  hugetlb.2MB.rsvd.current  irq.pressure    memory.peak          misc.current
cgroup.subtree_control  cpuset.mems            hugetlb.1GB.max           hugetlb.2MB.rsvd.max      memory.current  memory.pressure      misc.events
$ cat /sys/fs/cgroup/system.slice/run-u4260.scope/memory.max 
262144000
```

(basically, the resources and cgroups are switched in the hierarchy).

Reply:
I'm on Ubuntu 20.04.... Thanks for the info!

Reply:
Just tested on Ubuntu 22.04, it's using the same (v2) hierarchy.

Reply:
support for cgroup v1 (tested on 20.04) and v2 (tested on 22.04) added